tc_build/source.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+
+import hashlib
+import re
+import subprocess
+
+# Allows being imported via tc_build package or directly in REPL
+try:
+    import utils
+except ModuleNotFoundError:
+    from . import utils
+
+
+class Tarball:
+
+    def __init__(self):
+        self.base_download_url = None
+        self.local_location = None
+        self.remote_tarball_name = None
+        self.remote_checksum_name = ''
+
+    def download(self):
+        if not self.local_location:
+            raise RuntimeError('No local tarball location specified?')
+        if self.local_location.exists():
+            return  # Already downloaded
+
+        if not self.base_download_url:
+            raise RuntimeError('No tarball download URL specified?')
+        if not self.remote_tarball_name:
+            self.remote_tarball_name = self.local_location.name
+
+        utils.curl(f"{self.base_download_url}/{self.remote_tarball_name}",
+                   destination=self.local_location)
+
+        # If there is a remote checksum file, download it, find the checksum
+        # for the particular tarball, compute the downloaded file's checksum,
+        # and finally compare the two.
+        if self.remote_checksum_name:
+            checksums = utils.curl(f"{self.base_download_url}/{self.remote_checksum_name}")
+            if not (match := re.search(
+                    fr"([0-9a-f]+)\s+{self.remote_tarball_name}$", checksums, flags=re.M)):
+                raise RuntimeError(f"Could not find checksum for {self.remote_tarball_name}?")
+
+            if 'sha256' in self.remote_checksum_name:
+                file_hash = hashlib.sha256()
+            elif 'sha512' in self.remote_checksum_name:
+                file_hash = hashlib.sha512()
+            else:
+                raise RuntimeError(
+                    f"No supported hashlib for {self.remote_checksum_name}, add support for it?")
+            with self.local_location.open('rb') as file:
+                while (data := file.read(131072)):
```

Comment:
That's a magical constant number.  Why don't you give it a meaningful identifier?

Reply:
Sure thing!

01b65b97a480b3eeeab0c6759d2bfef6836f7189